components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,216 @@
+  }
+
+  eligibility_observation_.Reset();
+}
+
+void LearningService::Init() {
+  DCHECK(url_loader_factory_);
+  DCHECK(eligibility_service_);
+  DCHECK(init_task_timer_);
+
+  VLOG(1) << "Initializing federated learning service.";
+  eligibility_observation_.Observe(eligibility_service_);
+  if (eligibility_service_->IsEligible()) {
+    StartParticipating();
+  }
+
+  VLOG(1) << "Eligibility: " << eligibility_service_->IsEligible();
+}
+
+void LearningService::OnEligibilityChanged(bool is_eligible) {
+  if (is_eligible) {
+    StartParticipating();
+    VLOG(2) << "Eligibility changed, started participating.";
+  } else {
+    StopParticipating();
+    VLOG(2) << "Eligibility changed, stopped participating.";
+  }
+}
+
+void LearningService::StartParticipating() {
+  if (participating_) {
+    return;
+  }
+
+  participating_ = true;
+  GetTasks();
+}
+
+void LearningService::StopParticipating() {
+  if (!participating_) {
+    return;
+  }
+
+  participating_ = false;
+  reconnect_timer_.reset();
+}
+
+void LearningService::GetTasks() {
+  communication_adapter_->GetTasks(base::BindOnce(
+      &LearningService::HandleTasksOrReconnect, weak_factory_.GetWeakPtr()));
+}
+
+void LearningService::HandleTasksOrReconnect(TaskList tasks, int reconnect) {
+  if (tasks.size() == 0) {
+    reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    reconnect_timer_->Start(FROM_HERE, base::Seconds(reconnect), this,
+                            &LearningService::GetTasks);
+    VLOG(2) << "No tasks available, reconnecting in " << reconnect << "s";
+    return;
+  }
+
+  Task task = tasks.at(0);
+  float lr = 0.01;
+  std::map<std::string, float> config = task.GetConfig();
+  auto cursor = config.find("lr");
+  if (cursor != config.end()) {
+    lr = cursor->second;
+    VLOG(2) << "Learning rate applied from server: " << lr;
+  }
+
+  model_spec_->learning_rate = lr;
+
+  if (static_cast<int>(task.GetParameters().at(0).size()) !=
+          model_spec_->num_params &&
+      task.GetParameters().at(1).size() != 1U) {
+    VLOG(2) << "Task specifies a different model size than the client";
+    return;
+  }
+  VLOG(2) << "Task model and client model match!";
+
+  std::unique_ptr<Model> model = std::make_unique<Model>(*model_spec_);
+  model->SetWeights(task.GetParameters().at(0));
+  model->SetBias(task.GetParameters().at(1).at(0));
+  auto task_runner =
+      std::make_unique<FederatedTaskRunner>(task, std::move(model));
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      {base::TaskPriority::BEST_EFFORT,
+       base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN},
+      base::BindOnce(&LoadDatasetAndRunTask, std::move(task_runner)),
+      base::BindOnce(&LearningService::OnTaskResultComputed,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void LearningService::OnTaskResultComputed(absl::optional<TaskResult> result) {
+  if (!result.has_value()) {
+    VLOG(2) << "Task result computation failed";
```

Comment:
it seems like there is a lot of logging when errors happen, but no actual error handling. Do we not care if this fails? Also wouldn't it be useful to know why it failed for debugging purposes?